### PR TITLE
Watersensor

### PR DESCRIPTION
**Pull request by emilybthorpe**

### CODE REVIEW

**Comment by emilybthorpe** [comment]
After switching to the smartfin that Jordan was testing with, all manufacturing tests pass

**Comment by ntlhui** [comment]
please update branch

**Comment by ntlhui** [comment]
Please resolve conflicts

File: src/watersensor/waterSensor.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+{
+    // increment array location
+    array_location = (array_location + 1) % water_detect_array_size;
+    // subtract last value in the window from the rolling sum
+    array_sum -= water_detect_array[waterDetectArrayLocation(array_location, (-1 * moving_window_size))];
+
+    digitalWrite(water_detect_en_pin, LOW);
+    delayMicroseconds(WATER_DETECT_EN_TIME_US);
+    water_detect_array[array_location] = digitalRead(water_detect_pin); // comment out if using above logic block
+    digitalWrite(water_detect_en_pin, HIGH);
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:53]: Note that current measurement on a PicoScope indicates a neg pulse width of 1.010 ms with 1.289 us standard deviation over 100 measurements.

File: src/watersensor/waterSensor.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    // add value to the current sum
+    array_sum += water_detect_array[array_location];
+
+    // increment sample counter if needed
+    if (samples_taken_since_reset < moving_window_size)
+        samples_taken_since_reset++;
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:60]: Add conditional body braces

File: src/watersensor/waterSensor.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+     * we compare to high_detect_percentage of the moving window size because if
+     * all of the samples measured so far are 1, this is sufficient to say we
+     * are in the water.
+     */
+    if (samples_taken_since_reset < moving_window_size / 100. * high_detect_percentage)
+    {
+        return last_water_detect;
+    }
+
+    // out of the water!
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:74]: This should go inside the condition body

File: src/watersensor/waterSensor.cpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    else if ((((uint16_t)array_sum * 100) / moving_window_size) < low_detect_percentage)
+    {
+        last_water_detect = 0;
+        return last_water_detect;
+    }
+
+    // in the water!
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:81]: This should go inside the condition body

File: src/watersensor/waterSensor.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    else if ((((uint16_t)array_sum * 100) / moving_window_size) > high_detect_percentage)
+    {
+        last_water_detect = 1;
+        return last_water_detect;
+    }
+
+    // nothing has changed, return current status
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:88]: This should go inside the condition body

File: src/watersensor/waterSensor.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+{
+    return this->water_detect_array[this->array_location];
+}
+
+uint8_t WaterSensor::getLastStatus()
+{
+    return last_water_detect;
+}
+
+// gets the current in/out of water status (return true = in water, false = out)
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:110]: Use Doxygen style

File: src/watersensor/waterSensor.cpp [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+uint8_t WaterSensor::getCurrentStatus()
+{
+    return last_water_detect;
+}
+
+// forces the in/out of water status.  Usefule because of hystersis
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:116]: Use Doxygen style

File: src/watersensor/waterSensor.cpp (hunk 1 above)
Comment (ntlhui) [reply src/watersensor/waterSensor.cpp:53]: This should use `WaterSensor::getCurrentReading`

File: src/watersensor/waterSensor.cpp [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    digitalWrite(water_detect_en_pin, LOW);
+    delayMicroseconds(WATER_DETECT_EN_TIME_US);
+    temp_8 = digitalRead(water_detect_pin); // comment out if using above logic block
+    digitalWrite(water_detect_en_pin, HIGH);
+
+    return temp_8; // comment out if using above logic block
+}
+
+// set the low detection percentage of the moving window for hystersis
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:137]: Use Doxygen style

File: src/watersensor/waterSensor.cpp [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+bool WaterSensor::setLowDetectPercentage(uint8_t low_percentage)
+{
+    if ((low_percentage >= 0) && (low_percentage <= 100))
+    {
+        low_detect_percentage = low_percentage;
+        return true;
+    }
+    else
+        return false;
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:146]: Add braces

File: src/watersensor/waterSensor.cpp [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+}
+// set the high detection percentage of the moving window for hystersis
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:148]: Use Doxygen style

File: src/watersensor/waterSensor.cpp [hunk 11]
```diff
@@ -0,0 +1,169 @@
+#include "waterSensor.hpp"
+#include "product.hpp" // Added by PJB. Is it conventional to do this? Not sure but we need USB_PWR_DETECT_PIN
+#include "cli/conio.hpp"
+
+WaterSensor::WaterSensor(uint8_t water_detect_en_pin_to_set, uint8_t water_detect_pin_to_set, uint8_t window_size, uint8_t array_size) : water_detect_en_pin(water_detect_en_pin_to_set), water_detect_pin(water_detect_pin_to_set), moving_window_size(window_size), water_detect_array_size(array_size)
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:5]: Split into multiple lines

File: src/watersensor/waterSensor.cpp [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+uint8_t WaterSensor::waterDetectArrayLocation(int16_t location, int16_t offset)
+// returns a valid location (between 0 and WATER_DETECT_ARRAY_SIZE) for a location + an offset
+{
+    if ((location + offset) < 0)
+        return (water_detect_array_size + (location + offset));
+    else if ((location + offset) >= water_detect_array_size)
+        return ((location + offset) - water_detect_array_size);
+    else
+        return (location + offset);
```
Comment (ntlhui) [inline src/watersensor/waterSensor.cpp:168]: Add braces

File: src/watersensor/waterSensor.hpp
Comment (ntlhui) [inline src/watersensor/waterSensor.hpp:1]: Update docs to Doxygen style

**Review by ntlhui — CHANGES_REQUESTED** [review]

File: src/mfgTest/mfgTest.hpp [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     * @brief 
+     * 
+     */
+    void init(void);
+    STATES_e run(void);
+    void exit(void);
+
+    static int wet_dry_sensor_test(void);
+    static int(* MFG_TEST_TABLE[])(void);
```
[the remark this answers is not in the document]
Comment (emilybthorpe) [reply src/mfgTest/mfgTest.hpp:20]: Probably not, going to mark as private

**Comment by ntlhui** [comment]
Fixes #8

**Review by ntlhui — APPROVED** [review]

**Pull request merged**